sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSessionCache.java

```diff
@@ -0,0 +1,264 @@
+     *
+     * @param name the name of the session to evict.
+     * @return true if the session was evicted, false if no session found with the given name.
+     */
+    boolean evict(String name) {
+        if (name == null) {
+            return false;
+        }
+        final Entry removed = entries.remove(name);
+        if (removed != null) {
+            removed.dispose();
+        }
+        return removed != null;
+    }
+
+    /**
+     * Signal that the owner ({@link ReactorConnection}) of the cache is disposed of.
+     */
+    void setOwnerDisposed() {
+        isOwnerDisposed.set(true);
+    }
+
+    /**
+     * When the owner {@link ReactorConnection} is being disposed of, all {@link ReactorSession} loaded into the cache
+     * will receive shutdown signal through the channel established at ReactorSession's construction time, the owner
+     * may use this method to waits for sessions to complete it closing.
+     *
+     * @return a Mono that completes when all sessions are closed via owner shutdown signaling.
+     */
+    Mono<Void> awaitClose() {
+        final ArrayList<Mono<Void>> closing = new ArrayList<>(entries.size());
+        for (Entry entry : entries.values()) {
+            closing.add(entry.awaitSessionClose());
+        }
+        return Mono.when(closing);
+    }
+
+    /**
+     * Load a new {@link ReactorSession} to be cached.
+     *
+     * @param connection the QPid Proton-j connection to host the session.
+     * @param name the session name.
+     * @param loader the function to load the session.
+     *
+     * @return the session to cache.
+     */
+    private ReactorSession load(Connection connection, String name, Loader loader) {
+        final ProtonSession protonSession = new ProtonSession(connectionId, fullyQualifiedNamespace, connection,
+            handlerProvider, reactorProvider, name, openTimeout, logger);
+        // TODO (anu): Update loader signature to use 'ProtonSession' instead of 'ProtonSessionWrapper' when removing v1.
+        return loader.load(new ProtonSessionWrapper(protonSession));
+    }
+
+    /**
+     * Register to evict the session from the cache when the session terminates.
+     *
+     * @param session the session to register for cache eviction.
+     * @return the registration disposable.
+     */
+    private Disposable setupAutoEviction(ReactorSession session) {
+        return session.getEndpointStates().subscribe(__ -> {
+        }, error -> {
+            evict(session, "Evicting session terminated with error.", error);
+        }, () -> {
+            evict(session, "Evicting terminated session.", null);
+        });
+    }
+
+    /**
+     * Attempt to evict the session from the cache.
+     *
+     * @param session the session to evict.
+     * @param message the message to log on eviction.
+     * @param error the error triggered the eviction.
+     */
+    private void evict(ReactorSession session, String message, Throwable error) {
```

Comment:
nit: i try to put all same named methods and overloads together so I know what options are there. I think the other evict method is up there.

Reply:
the reason for picking this current order was – the other _evict_ is package private and this one is private. So, I was grouping methods based on visibility, typically we follow such ordering for member methods, variables (like public -> package private -> private)

Reply:
resolving this based on the above comment. let me know if we need to revisit this.